smart_open/s3.py

```diff
@@ -807,7 +828,9 @@ def _upload_next_part(self):
         logger.debug("upload of part #%i finished" % part_num)
 
         self._total_parts += 1
-        self._buf = io.BytesIO()
+
+        self._buf.seek(0)
+        self._buf.truncate(0)
```

Comment:
This is interesting. Hypothetically, users could pass in another smart_open object as the buffer, right? As long as it supports `truncate(0)`.

Reply:
In theory, yes. I haven't gone done that particular inception-like rabbit hole yet, though.